### PR TITLE
fix: recentDocuments menu role on macOS

### PR DESCRIPTION
**Pull request by codebytere**
#### Description of Change

Fixes https://github.com/electron/electron/issues/20364.

In the past when we constructed our menu through `MainMenu.xib`, we got Chromium's submenu construction defaults for free, which is why prior to `v6.0.0-beta.1` the `recentDocuments` role worked as we intended. The `recentDocuments` NSMenuItem would never have a `nil` submenu, since barring any actual recent documents it had a default `Clear Menu` item (see `.xib` file in linked CL). When they removed their dependence on MainMenu.xib via  [this CL](https://chromium-review.googlesource.com/c/chromium/src/+/1627242) it had the unfortunate side-effect that [this call](https://github.com/electron/electron/blob/eb2d2264d02abd7f34df36809d44752aaf93da93/shell/browser/ui/cocoa/atom_menu_controller.mm#L196-L197) would always return a null pointer, and so `[item setSubmenu:recentDocumentsMenu]` would _never set a submenu_ and thus never make the NSMenuItem's `action` selector `submenuAction:`. As a result, `validateUserInterfaceItem` would return `NO` and the item would both have no submenu and also be mysteriously disabled. 

To fix this, we now allocate an `NSMenu` for recent documents and append the old default menu item to it if no submenu exists (ie we're appending it for the first time). I've also gone ahead and added a pair of DCHECKs to prevent a similar issue from occurring in the future.
 
cc @nornagon @jkleinsc @zcbenz 

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Fixed a regression in the `recentDocuments` MenuItem role on macOS.

### CODE REVIEW

File: shell/browser/ui/cocoa/atom_menu_controller.mm [hunk 1]
```diff
@@ -193,8 +193,18 @@ - (void)moveMenuItems:(NSMenu*)source to:(NSMenu*)destination {
...
-  NSMenu* recentDocumentsMenu =
-      [[[recentDocumentsMenuItem_ submenu] retain] autorelease];
+  NSMenu* recentDocumentsMenu = nil;
+  if ([recentDocumentsMenuItem_ submenu]) {
+    recentDocumentsMenu =
+        [[[recentDocumentsMenuItem_ submenu] retain] autorelease];
+  } else {
+    recentDocumentsMenu = [[NSMenu alloc] initWithTitle:@""];
+    NSMenuItem* clearItem =
+        [[NSMenuItem alloc] initWithTitle:@"Clear Menu"
```
Comment (nornagon) [inline shell/browser/ui/cocoa/atom_menu_controller.mm]: Does this need to be localized?

Comment (codebytere) [reply shell/browser/ui/cocoa/atom_menu_controller.mm]: not particularly :)

File: shell/browser/ui/cocoa/atom_menu_controller.mm [hunk 2]
```diff
@@ -193,8 +193,18 @@ - (void)moveMenuItems:(NSMenu*)source to:(NSMenu*)destination {
 // Replaces the item's submenu instance with the singleton recent documents
 // menu. Previously replaced menu items will be recovered.
 - (void)replaceSubmenuShowingRecentDocuments:(NSMenuItem*)item {
-  NSMenu* recentDocumentsMenu =
-      [[[recentDocumentsMenuItem_ submenu] retain] autorelease];
+  NSMenu* recentDocumentsMenu = nil;
+  if ([recentDocumentsMenuItem_ submenu]) {
+    recentDocumentsMenu =
+        [[[recentDocumentsMenuItem_ submenu] retain] autorelease];
```
Comment (zcbenz) [inline shell/browser/ui/cocoa/atom_menu_controller.mm]: `[[retain] autorelease]` is no-op, `recentDocumentsMenu = [recentDocumentsMenuItem_ submenu]` is enough.

File: shell/browser/ui/cocoa/atom_menu_controller.mm [hunk 3]
```diff
@@ -193,8 +193,18 @@ - (void)moveMenuItems:(NSMenu*)source to:(NSMenu*)destination {
...
+  } else {
+    recentDocumentsMenu = [[NSMenu alloc] initWithTitle:@""];
```
Comment (zcbenz) [inline shell/browser/ui/cocoa/atom_menu_controller.mm]: Should do `autorelease` here.

File: shell/browser/ui/cocoa/atom_menu_controller.mm [hunk 4]
```diff
@@ -193,8 +193,18 @@ - (void)moveMenuItems:(NSMenu*)source to:(NSMenu*)destination {
...
+    [recentDocumentsMenu
+        addItem:[[NSMenuItem alloc]
```
Comment (zcbenz) [inline shell/browser/ui/cocoa/atom_menu_controller.mm]: Should do `autorelease` for the menu item.

File: shell/browser/ui/cocoa/atom_menu_controller.mm (hunk 1 above)
Comment (nornagon) [reply shell/browser/ui/cocoa/atom_menu_controller.mm]: Will this show in English to users whose system language is set to a non-English language?

Comment (codebytere) [reply shell/browser/ui/cocoa/atom_menu_controller.mm]: ah great catch, ty 🙇‍♀

**Pull request merged**